### PR TITLE
change(state): Refactor the naming of note commitment subtrees

### LINKED ISSUE #7463: Rename subtree methods and types to "subtree root"
## Schedule

This cleanup should happen last, because it will conflict with almost every other subtree change.

## Motivation

We are storing subtree roots and end heights in the subtree indexes, not the subtree trees themselves.

This looks like the sort of information that gets lost and causes misunderstandings in future :) Can we rename the methods and update comments to correctly reflect what we are storing? Since we want this PR to merge soon we should do it in a follow up PR

_Originally posted by @mpguerra in https://github.com/ZcashFoundation/zebra/pull/7437#discussion_r1314853877_
            

### Suggested Changes

Open 2 PRs, one for mass renames, and one for comments.

#### Mass Renames

- Rename types
- Rename methods and their arguments

Make a list of renames and then do a mass rename:
https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/mass-renames.md#universal-renames-with-sed

Don't rename the database column families, because that requires a state upgrade, and they are technically correct (every value is an index to a subtree):
https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/state-db-upgrades.md#current-state-database-format

#### Comments

Update the comments of each subtree type and method to make it clear that the subtree indexes contain subtree heights and roots.

### PR DESCRIPTION
**Pull request by upbqdn**
## Motivation

Close #7463.

## Solution

#7463 mentions that "_We are storing subtree roots and end heights in the subtree indexes, not the subtree trees themselves._" I thought of renaming `NoteCommitmentSubtree` to `SubtreeRoot`, but there were at least two problems with that:

1. There is also the struct `NoteCommitmentSubtreeData`, which I wasn't sure how to rename.
2. `NoteCommitmentSubtree` has a field `node`, which I had renamed to `root`, so we would end up with something like `subtree_root.root`.

Given the issues above, I decided to do a simpler refactor:

1. Rename the generic type `Node` in `NoteCommitmentSubtree` to `Root`.
2. Rename the field `node` in `NoteCommitmentSubtree` to `root`.
3. Rename the field `end` in `NoteCommitmentSubtree` to `end_height` to match the name in the RPC so that the field doesn't need to be renamed for the RPC.

After this refactor, I didn't find any docs that would require changes. 

Also, if we ever need to store whole subtrees instead of only roots, we can use the already existing struct `NoteCommitmentSubtree`, but keep the RPC that currently uses this struct intact because it actually uses `NoteCommitmentSubtreeData`. 

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

### CODE REVIEW

**Comment by upbqdn** [comment]
I'm not actually sure about the solution in this PR. Do others have suggestions for better solutions?

**Review by arya2 — DISMISSED** [review]
Looks good, thank you!

File: zebra-chain/src/subtree.rs [hunk 1]
```diff
@@ -50,61 +50,64 @@ impl From<NoteCommitmentSubtreeIndex> for u64 {
...
 // - consider defining sapling::SubtreeRoot and orchard::SubtreeRoot types or type wrappers,
 //   to avoid type confusion between the leaf Node and subtree root types.
-// - rename the `Node` generic to `SubtreeRoot`
 
 /// Subtree root of Sapling or Orchard note commitment tree,
 /// with its associated block height and subtree index.
 #[derive(Copy, Clone, Debug, Eq, PartialEq)]
 #[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
-pub struct NoteCommitmentSubtree<Node> {
+pub struct NoteCommitmentSubtree<Root> {
```
Comment (arya2) [inline zebra-chain/src/subtree.rs]: Optional/Nitpick: Rename this to `SubtreeRoot` to avoid confusion with other sapling/orchard `Root` types.

Comment (upbqdn) [reply zebra-chain/src/subtree.rs]: Done in [a278777](https://github.com/ZcashFoundation/zebra/pull/7855/commits/a278777decfe04dfc265badd6c1049ab2697b83e).

**Review by arya2 — APPROVED** [review]
Thank you!

**Pull request merged**